### ISSUE TITLE
A Couple Bugs in Set syntax

### ISSUE THREAD

**Issue by sidbav** [olofhagsand/cligen#34]
Given the following syntax:
```
aa @{
    bb<int32>;
    cc;
    dd;
}
```

If i type in the following, I see the following:
```
cli > aa bb <TAB>
 cc                    dd                    <int32>
```

This should instead produce:
```
cli > aa bb <TAB>
 <int32>
```

Another bug I have found is associated with the following syntax:
```
aa @{
    bb;
    cc;
    dd;
}
aa ff;
```
If I type in the following and then hit tab, I see: 
```
cli> aa ff <TAB>
 bb                    cc                    dd  
```
In the case I should not see anything when hitting tab. It should look like this:
```
cli> aa ff <TAB>
cli> aa ff 
```

**Comment by olofhagsand** [comment]
Actually, the first case is the way I choose to interpret "set". Example:
aa @{ 
  bb {
    subtree;
  }
  cc;
}

will produce: aa bb cc; aa cc bb; AND aa bb subtree;
That is, after aa bb ? both "cc" and "subtree" should be possible.
[substitute "subtree" for <int32>].
The alternative would be:
aa bb cc; aa cc bb; AND aa bb subtree cc
Is that what you had expected? 
But "subtree" can be any complex sub-structure, such as { x;y; } or  even a @tree recursion, for example, which makes this very complex.

**Comment by olofhagsand** [comment]
Fixed nr 2 above

**Comment by sidbav** [comment]
Confirmed nr 2 works now

**Comment by sidbav** [comment]
The behavior I was expecting was:
aa bb subtree cc AND aa cc bb subtree. 

If you were to type in bb, then I believe you should its subtree, before moving on to nodes that on the same level as bb.

**Comment by olofhagsand** [comment]
Need some more thought.

**Comment by olofhagsand** [comment]
For the time being disabled USE_SETS by default. Enable by setting USE_SETS in cligen_custom.h

**Comment by owenlejeune** [comment]
What is the current state of this?  If the intention is to treat variables the same as subtrees (eg.
```
aa @{
    bb<int32>;
    cc;
}
```
expanding to `aa bb <int32>` and `aa bb cc`) , is there anyway we could do a custom implementation which would only allow `cc` to be entered after the variable `<int32>` is entered?  This is currently very high priority for us.

**Comment by olofhagsand** [comment]
The status is that the "set" feature is experimental. 
It is only applicable to single keywords. such as: `@{a;b;}`
It can not be used with complex sub-structures, such as:  `@{a <string>;b{c;} }` which leads to the (erroneous) behavior you describe above.
After investigating it in some detail I dont think it can be solved with a simple extension, it needs a larger restructuring of the current CLIgen design where higher level syntax (eg [], |, and @{]) is just mapped to the underlying command-based structure. 
In other words, more advanced "set", "list", etc, need to be first class objects, which would require more work than can be done in a couple of hours, or even days.
I realize that this would be useful, but I am constrained by time.
BTW, who do you represent (you say "us")?

**Comment by owenlejeune** [comment]
Thanks for the update.  We will have to do some more discussion ourselves on the best way for us to move forward.

I am currently a co-op with Ciena.

**Comment by owenlejeune** [comment]
Sorry to push you on this again, but are there any plans on the roadmap to extend the "sets" feature functionality?  Using sets with complex sub-structures (mostly keywords with associated variables) is a requirement for our project.  If there are no plans to implement this, we will have to investigate implementing a solution on our own.

**Comment by olofhagsand** [comment]
It is not on any planned roadmap. I am primarily focusing on the clixon roadmap, where cligen is only as a support funtion.
I realize "the set" functionality is important for many, but there is no easy fix, and I will look at it when I have time. That's all I can say presently, sorry.
Can we continue this discussion on direct email?

**Comment by owenlejeune** [comment]
I appreciate the update.  If I could email you directly that would be great, what email address can I reach you at?

**Comment by olofhagsand** [comment]
See http://cligen.se/
